apollo-router/src/plugins/connectors/http_json_transport.rs

```diff
@@ -77,43 +78,72 @@ pub(crate) fn make_request(
+        .as_ref()
+        .map(|v| v.to_str().unwrap_or_default())
+        == Some("application/x-www-form-urlencoded");
+
+    let (json_body, form_body, body, apply_to_errors) = if let Some(ref selection) = transport.body
+    {
         let (json_body, apply_to_errors) = selection.apply_with_vars(&json!({}), &inputs);
+        let mut form_body = None;
         let body = if let Some(json_body) = json_body.as_ref() {
-            hyper::Body::from(serde_json::to_vec(json_body)?)
+            if is_form_urlencoded {
+                let encoded = encode_json_as_form(json_body.as_object().unwrap());
+                form_body = Some(encoded.clone());
+                hyper::Body::from(encoded)
+            } else {
+                request = request.header(CONTENT_TYPE, "application/json");
+                hyper::Body::from(serde_json::to_vec(json_body)?)
+            }
         } else {
             hyper::Body::empty()
         };
-        (json_body, body, apply_to_errors)
+        (json_body, form_body, body, apply_to_errors)
     } else {
-        (None, hyper::Body::empty(), vec![])
+        (None, None, hyper::Body::empty(), vec![])
     };
 
-    let mut request = http::Request::builder()
-        .method(transport.method.as_str())
-        .uri(uri.as_str())
-        .header("content-type", "application/json")
+    let request = request
         .body(body.into())
         .map_err(HttpJsonTransportError::InvalidNewRequest)?;
 
-    add_headers(
-        &mut request,
-        original_request.supergraph_request.headers(),
-        &transport.headers,
-        &flat_inputs,
-    );
-
     if let Some(debug) = debug {
-        debug.lock().push_request(
-            &request,
-            json_body.as_ref(),
-            transport.body.as_ref().map(|body| SelectionData {
-                source: body.to_string(),
-                transformed: body.to_string(),
-                result: json_body.clone(),
-                errors: apply_to_errors,
-            }),
-        );
+        if is_form_urlencoded {
+            debug.lock().push_form_urlencoded_request(
+                &request,
+                form_body.as_ref(),
+                transport.body.as_ref().map(|body| SelectionData {
+                    source: body.to_string(),
+                    transformed: body.to_string(), // no transformation so this is the same
+                    result: json_body,
+                    errors: apply_to_errors,
+                }),
+            );
+        } else {
+            debug.lock().push_request(
+                &request,
+                json_body.as_ref(),
+                transport.body.as_ref().map(|body| SelectionData {
+                    source: body.to_string(),
+                    transformed: body.to_string(), // no transformation so this is the same
+                    result: json_body.clone(),
+                    errors: apply_to_errors,
+                }),
+            );
```

Comment:
I think the only differences between these two, down in the new functions, is that the JSON body is converted to a string later than the form body and that "kind" is set. So can we consolidate the functions if we pass in `kind` here and do the json to_string here?

Reply:
yeah probably! i'll look into it

Reply:
done!